### DISCUSSION TITLE
pods get recreated daily

### DISCUSSION THREAD

**Discussion by mkreuzmayr** [kube-hetzner/terraform-hcloud-kube-hetzner#466] [Show and tell]
I don´t know if this has to do someting with kube-hetzner in particular but I noticed the following today and I cannot think of what happened. Somehow every day all of my pods get recreated with the exception of some in the pods kube-system namespace, they only get restarted.

![kube](https://user-images.githubusercontent.com/20108212/208396093-4ec47538-8743-42d2-ad8d-871d9f092596.jpg)

I have got three nodes and they are all up since 25d. My hetzner control panel also does not show any activities concerning node creation/deletion. It only shows service updates and volume/attach detaches.

![events](https://user-images.githubusercontent.com/20108212/208400012-2e5a0c8b-7c0d-4ec4-8737-265a0ee380f7.jpg)

I already inspected logs and events in kubernetes or if I have any jobs or tasks that could lead to that.

**Comment by mrhein** [comment]
EDITED

in short: microos tumbleweed means daily updates  + kured => autopatching

microos tumbleweed means daily updates => https://ftp.gwdg.de/pub/opensuse/tumbleweed/appliances/
you can see the changes here: https://ftp.gwdg.de/pub/opensuse/tumbleweed/iso/

**Comment by mkreuzmayr** [comment]
Ok, that makes sense. If i have multiple replicas of my pods, will the service be available while updating or is there a short downtime?

**Comment by otavio** [comment]
This also concerns me. We are moving our services to this and the downtime is a bad thing for us.

**Comment by mrhein** [comment]
hopefully it should not be problem otherwise you need to check why it is a problem for your services in best practise world it shouldn't be a problem because only one node after one is rebooting

dDatabases (Postgres) also run without problems
![image](https://user-images.githubusercontent.com/1809310/208499134-a06c5a3a-5b31-4e5e-b14e-63c3b5c4120b.png)
![image](https://user-images.githubusercontent.com/1809310/208499229-06922af6-5e81-4e29-9e22-a498ba8d6cf4.png)

**Comment by maaft** [comment]
I have the same issue, but in particular in combination with Hashicorp Vault (one of the most popular secret stores): After being restarted, the vault gets automatically sealed. And for security purposes, the unseal process must be manual.

Meaning, our corporate secret store (vault) would required manual intervention every day to stay functional.

Any idea what I can do instead? Could we do manual, scheduled upgrades of MicroOS instead? If so, how?

**edit** vault has an auto-unseal option. But it requires either AWS, GCP or Azure AD which I want to avoid. Or another self-hosted vault cluster. Chicken-Egg problem..

**p.s.** @mrhein What postgres operator are you using? Are you satisfied? I'm currently not sure if I'll go with bitnami or zalando.

**Comment by mrhein** [comment]
from architecture perspective you can solve it with a management cluster running rockylinux or simlar with your vault harbor and so on and your application cluster in a more updated way the problem is than you have 2 setups to manage and a more vulnerable platform 

maybe hetzner will plugin one of this YubiHSM 2 if we ask them (atleast for metal server), because vault says...
> Vault integrates with the YubiHSM 2 via PKS#11 for customers who need hardware backed security. The integration supports auto unseal, cryptographic key generation and management, & external entropy.

so or so the kured config should be edited
https://kured.dev/docs/configuration/#setting-a-schedule
https://github.com/kube-hetzner/terraform-hcloud-kube-hetzner/blob/master/kustomize/kured.yaml#L19

that should help you out maybe update times only 3 days in row in middle of the week 9 to 11 am

p.s. @maaft there are only one solution on the market that is great and that ist zalando => postgres opertor which spawns patroni => spilo for you

**Reply by maaft** [reply]
What exactly does make zalando operator "great" in your opinion? E.g. in comparison with the one from [bitnami](https://github.com/bitnami/charts/tree/main/bitnami/postgresql-ha)

Key differences from the top of my mind:
- zalando integrates wal-g for easy backups/recovery (for bitnami one has to figure out backups/restores themselves) 
- zalando uses patroni / spilo; bitnami uses `repmgr` ==> no idea which is "better"
- bitnami uses pgpool-II instead of pgbouncer, which I need anyway for R/W splitting so I'd need to install it anyway on top of zalando operator
- bitnami seems to release more frequently than zalando
- bitnami docs seem to be more polished

anything else?

**Reply by mysticaltech** [reply]
Hey folks, just FYI one of our initial contributors @mnencia is one of the lead architects of [cloudnative-pg.io](https://cloudnative-pg.io/), also an amazing pg operator I heard! :)

**Reply by maaft** [reply]
thx for that amazing shoutout! CNPG is a beast! <3

**Comment by mysticaltech** [comment]
@mrhein Exactly. This is by design. When an update comes, microOS applies in a new snapshot, then signals it to kured, which in turn drains the node and reboots to apply the new snapshot. If you are running in HA, >= 3 control planes, and > 2 agents, you should have 0 downtime.

Now @mkreuzmayr @maaft Note that you can disable this behavior if you choose so, there are variables for that, also see the manual way in the readme.

**Reply by maaft** [reply]
When we disable this behavior, how can we issue manual upgrades?

**Reply by mysticaltech** [reply]
So basically, you have two things.

Node upgrades, which is the behavior described above. And k3s upgrades are done by something called the system upgrade controller.

You can disable both and trigger both manually. Please search for "upgrade" or "update" in the readme, I've explained the process well there.

**Reply by mysticaltech** [reply]
In the readme it's the disable part. To do the actual upgrade of the node manually, you just do this.

1/ SSH into the node.
2/ transactional-update pkg upgrade
2/ touch /var/run/reboot-required

And wait for the magic to happen! Now you may need to change the kured config too in the kube.tf, so that it looks for that file every 5 minutes instead of every hour.

**Reply by mysticaltech** [reply]
For the system upgrade controller, even simpler, all can be done via yaml and kubectl applies. Just read the corresponding [doc](https://github.com/rancher/system-upgrade-controller) (and the readme that explains where the upgrade plans live).

**Reply by maaft** [reply]
thanks a bunch! I'll go with this implementation then.

**Reply by maaft** [reply]
@mysticaltech coming back to this as I just tried OS upgrades.

`transactional-update pkg upgrade` does not exist. There is `update` - but that is for individual packages only.. Any idea how to do it instead?

Edit: there is `transactional-update up` - might this be the chosen command?

**Reply by maaft** [reply]
Also, how can I change kured config via kube.tf? Didn't find any comments there..

**Reply by mysticaltech** [reply]
@maaft Yes, as per these [docs](https://kubic.opensuse.org/documentation/man-pages/transactional-update.8.html), it's either `transactional-update up` or `transactional-update dup`, honestly am not sure about the difference.

**Reply by mysticaltech** [reply]
@maaft For an already running cluster, just use kubectl to grab the kured daemonset, with `kubectl get daemonset kured -n kube-system -o yaml`. And change it to what you want! Like changing the period to with '- --period=60s', instead of the default of 1h. Then apply again!

**Reply by maaft** [reply]
From comments on the internet I conclude that I need to use `transactional-update dup`. But the reason is still unknown for me.

Edit:
As per  [this](https://de.opensuse.org/SDB:Zypper_benutzen) source (german) `dup` might be the prefered method for tumbleweed as it also does a "dist" upgrade. This might be the `d` in `dup`.

**Reply by mysticaltech** [reply]
Wonderful! Yes, even so it's a rolling release, it would make sense to use dup to make sure that things are updated to the maximum. Now, we can also have a look at a systemctl timer to see which one the automatic upgrade uses.

**Reply by maaft** [reply]
@mysticaltech I've got one issue though, maybe you can help.

After doing these steps:
```sh
transactional-update dup
touch /var/run/reboot-requried
```

kured successfully, drained the node and restarted it.

Afterwards, the nodes still are labeled with `kured=rebooting`. This prevents to use the `system-upgrade-controller` with the default upgrade plans. Any idea why the label is not removed/changed after reboot?

**Reply by maaft** [reply]
Ah, I think the kured daemonset misses the `--post-reboot-node-labels=kured=done` argument (or similar). Makes me wonder though, why with this worked in the first place with enabled upgrades via `kube.tf`

**Reply by maaft** [reply]
.... or *did it actually worked* before? Did you confirm functional automatic k3s upgrades?

As per the the deployed upgrade plan, nodes will not get selected when they have `kured=rebooting` label:
```yaml
apiVersion: upgrade.cattle.io/v1
kind: Plan
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"upgrade.cattle.io/v1","kind":"Plan","metadata":{"annotations":{},"labels":{"k3s_upgrade":"server"},"name":"k3s-server","namespace":"system-upgrade"},"spec":{"channel":"https://update.k3s.io/v1-release/channels/v1.24","concurrency":1,"cordon":true,"nodeSelector":{"matchExpressions":[{"key":"k3s_upgrade","operator":"Exists"},{"key":"k3s_upgrade","operator":"NotIn","values":["disabled","false"]},{"key":"node-role.kubernetes.io/control-plane","operator":"In","values":["true"]},{"key":"kured","operator":"NotIn","values":["rebooting"]}]},"serviceAccountName":"system-upgrade","tolerations":[{"effect":"NoSchedule","key":"node-role.kubernetes.io/control-plane","operator":"Exists"},{"effect":"NoExecute","key":"CriticalAddonsOnly","operator":"Exists"}],"upgrade":{"image":"rancher/k3s-upgrade"}}}
  creationTimestamp: "2022-12-20T13:13:15Z"
  generation: 4
  labels:
    k3s_upgrade: server
  name: k3s-server
  namespace: system-upgrade
  resourceVersion: "13865998"
  uid: ee5d1cc7-5a87-4659-99cc-3f8e60ee7f4b
spec:
  channel: https://update.k3s.io/v1-release/channels/v1.25
  concurrency: 1
  cordon: true
  nodeSelector:
    matchExpressions:
    - key: k3s_upgrade
      operator: Exists
    - key: k3s_upgrade
      operator: NotIn
      values:
      - disabled
      - "false"
    - key: node-role.kubernetes.io/control-plane
      operator: In
      values:
      - "true"
    - key: kured
      operator: NotIn
      values:
      - rebooting
  serviceAccountName: system-upgrade
  tolerations:
  - effect: NoSchedule
    key: node-role.kubernetes.io/control-plane
    operator: Exists
  - effect: NoExecute
    key: CriticalAddonsOnly
    operator: Exists
  upgrade:
    image: rancher/k3s-upgrade
status:
  conditions:
  - lastUpdateTime: "2023-01-10T11:13:17Z"
    reason: Channel
    status: "True"
    type: LatestResolved
  latestHash: 07e384d7b99896e9bfa9991cd9b6fbecc7a6de63057b51adbc35972f
  latestVersion: v1.25.4-k3s1
```

But the kured daemon set config does not change that label after boot (missing the --post-reboot-node-labels=kured=done`):
```yaml
apiVersion: apps/v1
kind: DaemonSet
metadata:
  annotations:
    deprecated.daemonset.template.generation: "2"
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"apps/v1","kind":"DaemonSet","metadata":{"annotations":{},"name":"kured","namespace":"kube-system"},"spec":{"selector":{"matchLabels":{"name":"kured"}},"template":{"metadata":{"labels":{"name":"kured"}},"spec":{"containers":[{"command":["/usr/bin/kured","--reboot-command=/usr/bin/systemctl reboot","--pre-reboot-node-labels=kured=rebooting"],"env":[{"name":"KURED_NODE_ID","valueFrom":{"fieldRef":{"fieldPath":"spec.nodeName"}}}],"image":"ghcr.io/kubereboot/kured:1.12.0","imagePullPolicy":"IfNotPresent","name":"kured","securityContext":{"privileged":true}}],"hostPID":true,"restartPolicy":"Always","serviceAccountName":"kured","tolerations":[{"effect":"NoSchedule","key":"node-role.kubernetes.io/control-plane"},{"effect":"NoSchedule","key":"node-role.kubernetes.io/master"}]}},"updateStrategy":{"type":"RollingUpdate"}}}
  creationTimestamp: "2022-12-20T13:12:46Z"
  generation: 2
  name: kured
  namespace: kube-system
  resourceVersion: "13823076"
  uid: f72b154b-6ff7-423f-a8ef-fde27b974b34
spec:
  revisionHistoryLimit: 10
  selector:
    matchLabels:
      name: kured
  template:
    metadata:
      creationTimestamp: null
      labels:
        name: kured
    spec:
      containers:
      - command:
        - /usr/bin/kured
        - --reboot-command=/usr/bin/systemctl reboot
        - --pre-reboot-node-labels=kured=rebooting
        - --period=60s
        env:
        - name: KURED_NODE_ID
          valueFrom:
            fieldRef:
              apiVersion: v1
              fieldPath: spec.nodeName
        image: ghcr.io/kubereboot/kured:1.12.0
        imagePullPolicy: IfNotPresent
        name: kured
        resources: {}
        securityContext:
          privileged: true
        terminationMessagePath: /dev/termination-log
        terminationMessagePolicy: File
      dnsPolicy: ClusterFirst
      hostPID: true
      restartPolicy: Always
      schedulerName: default-scheduler
      securityContext: {}
      serviceAccount: kured
      serviceAccountName: kured
      terminationGracePeriodSeconds: 30
      tolerations:
      - effect: NoSchedule
        key: node-role.kubernetes.io/control-plane
      - effect: NoSchedule
        key: node-role.kubernetes.io/master
  updateStrategy:
    rollingUpdate:
      maxSurge: 0
      maxUnavailable: 1
    type: RollingUpdate
status:
  currentNumberScheduled: 6
  desiredNumberScheduled: 6
  numberAvailable: 6
  numberMisscheduled: 0
  numberReady: 6
  observedGeneration: 2
  updatedNumberScheduled: 6
```

kubectl get nodes --show-labels | grep kured
```
cluster-cp-large-thj      Ready    control-plane,etcd,master   20d   v1.24.8+k3s1   beta.kubernetes.io/arch=amd64,beta.kubernetes.io/instance-type=cpx31,beta.kubernetes.io/os=linux,csi.hetzner.cloud/location=fsn1,egress.k3s.io/cluster=true,failure-domain.beta.kubernetes.io/region=fsn1,failure-domain.beta.kubernetes.io/zone=fsn1-dc14,kubernetes.io/arch=amd64,kubernetes.io/hostname=mgmt-cluster-cp-large-thj,kubernetes.io/os=linux,kured=rebooting,node-role.kubernetes.io/control-plane=true,node-role.kubernetes.io/etcd=true,node-role.kubernetes.io/master=true,node.kubernetes.io/exclude-from-external-load-balancers=true,node.kubernetes.io/instance-type=cpx31,topology.kubernetes.io/region=fsn1,topology.kubernetes.io/zone=fsn1-dc14

```

**Reply by maaft** [reply]
As soon as I changed that label manually, the k3s upgrade triggered.

@mysticaltech Might this be a general bug where we need a new issue? (automatic k3s upgrades not working when os upgrades are active)

**Reply by mysticaltech** [reply]
@maaft It makes sense... Damn! PR welcome for this urgent fix, please.

**Reply by Lennix** [reply]
Any updates to this?
I'm actually also trying to limit the upgrade schedule to once a week (during the weekend preferably) and reading this discussion made me look at my cluster and also notice that all my nodes have the kured=rebooting label...

**Reply by aleksasiriski** [reply]
> Any updates to this? I'm actually also trying to limit the upgrade schedule to once a week (during the weekend preferably) and reading this discussion made me look at my cluster and also notice that all my nodes have the kured=rebooting label...

There's an example in kube.tf

```terraform
  # If you need more control over kured and the reboot behaviour, you can pass additional options to kured.
  # For example limiting reboots to certain timeframes. For all options see: https://kured.dev/docs/configuration/
  # The default options are: `--reboot-command=/usr/bin/systemctl reboot --pre-reboot-node-labels=kured=rebooting --post-reboot-node-labels=kured=done --period=5m`
  # Defaults can be overridden by using the same key.
  kured_options = {
    "reboot-days": "su"
    "start-time": "3am"
    "end-time": "8am"
    "time-zone": "Local"
  }
```

If you want it during the whole weekend (if you have a lot of servers) set reboot-days to sa,su and remove other options. I personally set it to monday in the middle of the night

**Reply by mysticaltech** [reply]
@Lennix This has been fixed for a long time. But one of our oldest versions had a bug with the labels. You can fix it with the following procedure, which is quite simple https://github.com/kube-hetzner/terraform-hcloud-kube-hetzner/issues/562#issuecomment-1412653699. 

To limit once a week, @aleksasiriski provided sound advice above. Because updates do happen but are saved in a new btrfs snapshot, so they only kick in upon reboot!

**Comment by mkreuzmayr** [comment]
@mysticaltech I have got 3 control planes with control plane scheduling enabled, will it still be HA?

**Reply by mysticaltech** [reply]
@mkreuzmayr Most definitely!

**Reply by mysticaltech** [reply]
If your cluster is not resource-strained with the 3 nodes up and can handle one node going down for a few minutes, you are HA!
